### PR TITLE
Some work on the CUDA tutorial

### PR DESCRIPTION
**Pull request by masterleinad**
Add some in-code documentation and a results section.
Also eliminate a distributed host vector and fix some MPI issues. I am running into an issue with `dealii::internal::VectorOperations::functions<Number, Number, dealii::MemorySpace::CUDA>::dot` with two MPI processes, but am not sure that this is related to my CUDA-aware MPI setup, yet.

### CODE REVIEW

File: examples/step-85/doc/results.dox [hunk 1]
```diff
@@ -1 +1,10 @@
 <h1>Results</h1>
+
+<a name="extensions"></a>
+<h3> Possible extensions </h3>
+
+Currently, there is no preconditioner used at all. This is mainly since 
+constructing an efficient matrix-free preconditioner is non-trivial.
+However, simple choices just requiring the diagonal of the corresponding matrix
+are good candidates. In particular, one good extend the tutorial to use multigird
```
Comment (Rombur) [inline examples/step-85/doc/results.dox]: good -> could

File: examples/step-85/step-85.cu [hunk 2]
```diff
@@ -50,6 +51,13 @@ namespace Step85
 {
   using namespace dealii;
 
+
+  // Define a class that implements the varying coefficients we want to use in
+  // the Helmholtzoperator.
+  // Later, we wan't to pass an object of this type to a
```
Comment (Rombur) [inline examples/step-85/step-85.cu]: wan't -> want

File: examples/step-85/step-85.cu [hunk 3]
```diff
@@ -58,10 +66,18 @@ namespace Step85
       : coef(coefficient)
     {}
 
+    // Later, we wan't to pass an object of this type to a
```
Comment (Rombur) [inline examples/step-85/step-85.cu]: wan't -> want Also did you mean to have the same text copy and paste a few lines below

File: examples/step-85/step-85.cu [hunk 4]
```diff
@@ -125,7 +148,9 @@ namespace Step85
   }
 
 
-
+  // Finally, we need to define a class that implements the whole operator
+  // evaluation that corresponds to matrix-vector product in matr0x-based
```
Comment (Rombur) [inline examples/step-85/step-85.cu]: typo matrix

File: examples/step-85/step-85.cu [hunk 5]
```diff
@@ -141,6 +166,9 @@ namespace Step85
       const double *                                              src,
       double *                                                    dst) const;
 
+    // Again, the CUDAWrappers::MatrixFree object doesn't know about the number
+    // of degrees of freedoms and the number of degrees of freedoms so we need
```
Comment (Rombur) [inline examples/step-85/step-85.cu]: typo: number of degrees of freedoms and the number of degrees of freedoms

File: examples/step-85/step-85.cu [hunk 6]
```diff
@@ -218,7 +246,10 @@ namespace Step85
   }
 
 
-
+  // When applying the Helmholtz operator, we have to be carefull to handle
```
Comment (Rombur) [inline examples/step-85/step-85.cu]: typo careful

File: examples/step-85/step-85.cu [hunk 7]
```diff
@@ -218,7 +246,10 @@ namespace Step85
...
+  // boundary conditions correctly. Since the local operator doesn't knaow about
```
Comment (Rombur) [inline examples/step-85/step-85.cu]: typo know

File: include/deal.II/matrix_free/cuda_matrix_free.templates.h [hunk 8]
```diff
@@ -871,6 +871,9 @@ namespace CUDAWrappers
       return internal::get_conflict_indices<dim, Number>(filter, constraints);
     };
 
+    if (begin == end)
+      return;
+
```
Comment (Rombur) [inline include/deal.II/matrix_free/cuda_matrix_free.templates.h]: Is this meant to be in this PR?

Comment (masterleinad) [reply include/deal.II/matrix_free/cuda_matrix_free.templates.h]: Yes, `GraphColoring::make_graph_coloring` complains otherwise.

**Comment by masterleinad** [comment]
The current error message for running with two MPI processes is:
```
--------------------------------------------------------
An error occurred in line <2367> of file </export/home/darndt/dealii/include/deal.II/lac/vector_operations_internal.h> in function
    static Number dealii::internal::VectorOperations::functions<Number, Number, dealii::MemorySpace::CUDA>::dot(const std::shared_ptr<dealii::parallel::internal::TBBPartitioner>&, dealii::internal::VectorOperations::size_type, const dealii::MemorySpace::MemorySpaceData<Number, dealii::MemorySpace::CUDA>&, dealii::MemorySpace::MemorySpaceData<Number, dealii::MemorySpace::CUDA>&) [with Number = double; dealii::internal::VectorOperations::size_type = long long unsigned int]
The violated condition was: 
    error_code == cudaSuccess
Additional information: 
    invalid argument

Stacktrace:
-----------
#0  /export/home/darndt/dealii/build/lib/libdeal_II.g.so.9.1.0-pre: dealii::internal::VectorOperations::functions<double, double, dealii::MemorySpace::CUDA>::dot(std::shared_ptr<dealii::parallel::internal::TBBPartitioner> const&, unsigned long long, dealii::MemorySpace::MemorySpaceData<double, dealii::MemorySpace::CUDA> const&, dealii::MemorySpace::MemorySpaceData<double, dealii::MemorySpace::CUDA>&)
#1  /export/home/darndt/dealii/build/lib/libdeal_II.g.so.9.1.0-pre: void dealii::internal::VectorOperations::functions<double, double, dealii::MemorySpace::CUDA>::norm_2<double>(std::shared_ptr<dealii::parallel::internal::TBBPartitioner> const&, unsigned long long, double&, dealii::MemorySpace::MemorySpaceData<double, dealii::MemorySpace::CUDA>&)
#2  /export/home/darndt/dealii/build/lib/libdeal_II.g.so.9.1.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::CUDA>::norm_sqr_local() const
#3  /export/home/darndt/dealii/build/lib/libdeal_II.g.so.9.1.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::CUDA>::norm_sqr() const
#4  /export/home/darndt/dealii/build/lib/libdeal_II.g.so.9.1.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::CUDA>::l2_norm() const
#5  ./step-85: Step85::HelmholtzProblem<3, 3>::solve()
#6  ./step-85: Step85::HelmholtzProblem<3, 3>::run()
#7  ./step-85: main
```

**Comment by Rombur** [comment]
Yeah the memset looks wrong. It should be `size * sizeof(Number)` instead of `sizeof(Number)`

**Comment by Rombur** [comment]
Also it shouldn't be `Number()` but `0`. `cudaMemset` takes an integer.

**Comment by masterleinad** [comment]
Now, I am getting
```
An error occurred in line <2381> of file </export/home/darndt/dealii/include/deal.II/lac/vector_operations_internal.h> in function
    static Number dealii::internal::VectorOperations::functions<Number, Number, dealii::MemorySpace::CUDA>::dot(const std::shared_ptr<dealii::parallel::internal::TBBPartitioner>&, dealii::internal::VectorOperations::size_type, const dealii::MemorySpace::MemorySpaceData<Number, dealii::MemorySpace::CUDA>&, dealii::MemorySpace::MemorySpaceData<Number, dealii::MemorySpace::CUDA>&) [with Number = double; dealii::internal::VectorOperations::size_type = long long unsigned int]
The violated condition was: 
    cudaGetLastError() == cudaSuccess
Additional information: 
    no error
```
That looks weird. There is an error but the string is saying there is no error.

**Comment by Rombur** [comment]
I think that it should probably be unknown error instead of no error

**Comment by Rombur** [comment]
You are using cuda aware MPI right? Can you try to put a `cudaDeviceSynchronize` after the MPI communication. I wonder if the problem is that the MPI communication is not over when we get to the dot product.

**Comment by Rombur** [comment]
I bet it's the same error than we have [here](https://github.com/dealii/dealii/issues/7713)

**Comment by masterleinad** [comment]
This appears to finally work. We were using a too small vector for the coefficient values. We need one value for every quadrature point in each locally owned cell, i.e. `n_owned_cells * (q_points_1d)^dim`. This is not the same as the number of degrees of freedoms for continuous elements.

**Comment by Rombur** [comment]
There is a problem with the PR you need to rebase.

**Pull request merged**